### PR TITLE
Explicit source options overwrite

### PR DESCRIPTION
**Pull request by jingsam**
Explicit source options should taken precedence over TileJSON. See details in #8192

# Launch Checklist

 - [ ] briefly describe the changes in this PR
 - [ ] write tests for all new functionality
 - [ ] document any changes to public APIs
 - [ ] post benchmark scores
 - [ ] manually test the debug page
 - [ ] tagged `@mapbox/studio` and/or `@mapbox/maps-design` if this PR includes style spec changes
 - [ ] tagged `@mapbox/gl-native` if this PR includes shader changes or needs a native port

### CODE REVIEW

File: test/unit/style-spec/fixture/sources.output.json [hunk 1]
```diff
@@ -7,14 +7,6 @@
     "message": "sources.invalid-type.type: expected one of [vector, raster, raster-dem, geojson, video, image], \"invalid\" found",
     "line": 7
   },
-  {
-    "message": "sources.no-tilejson-properties-with-url.tiles: a source with a \"url\" property may not include a \"tiles\" property",
-    "line": 12
-  },
-  {
-    "message": "sources.no-unknown-properties-with-url.foo: a source with a \"url\" property may not include a \"foo\" property",
```
Comment (asheemmamoowala) [inline test/unit/style-spec/fixture/sources.output.json]: Why is unknown properties support needed?

Comment (jingsam) [reply test/unit/style-spec/fixture/sources.output.json]: I think foreign members (unknown properties) in source options should be leaved as it is. Foreign members does not affect source's functionality and add extensibility for metadata or custom sources. As we don't invalidate foreign members in geojson、image、video sources, invalidate foreign members in  vector, raster, raster-dem seems inconsistent.

**Comment by asheemmamoowala** [comment]
cc @mapbox/studio

File: test/unit/style-spec/fixture/sources.output.json (hunk 1 above)
Comment (asheemmamoowala) [reply test/unit/style-spec/fixture/sources.output.json]: @jingsam Sorry for the delay in getting back to this. Allowing foreign members in the TileJSON does not seem necessary as part of this PR. It is preferable to have strict validation to ensure that unnecesary or incorrect properties are not being added to such files.

**Comment by jingsam** [comment]
@asheemmamoowala I have invalid foreign members in source options when `url` is defined. As you can see, I have enum so many possible valid members. This code is fragile as someday we may need new property in test cases or application. If we now restrict properties allowed in sources options, that may make sources options can not be extended. Actually, in mapbox-style-spec, there is no limitation on foreign properties. So limit in here is sounds unreasonable. Forregin properties are hurtless, if mapbox-gl-js does not use them, take as it is but not restrict them.
https://github.com/mapbox/mapbox-gl-js/blob/58fa110a3360b2954d06af9ce43995631542586d/src/style-spec/validate/validate_source.js#L32-L38

**Comment by jingsam** [comment]
@asheemmamoowala I add a test, please check out.

**Pull request merged**